File: python/1137-N-th-Tribonacci-Number.py
class Solution:
    Memo = {}

    def tribonacci(self, n: int):
        if n in self.Memo:
            return self.Memo[n]
        if n == 0:
            return 0
        if n == 1:
            return 1
        if n == 2:
            return 1
        self.Memo[n] = (
            self.tribonacci(n - 1) + self.tribonacci(n - 2) + self.tribonacci(n - 3)
        )
        return self.Memo[n]

File: rust/1137-n-th-tribonacci-number.rs
impl Solution {
    pub fn tribonacci(n: i32) -> i32 {
        let mut t = [0, 1, 1];

        if n <= 2 {
            return t[n as usize];
        }

        for i in 3..(n + 1) as usize {
            t.swap(0, 1);
            t.swap(1, 2);
            t[2] = t.iter().sum();
        }

        t[2]
    }
}
